### PR TITLE
Feature/flow 3451 aes decryption

### PR DESCRIPTION
**Pull request by lmancel**
Here are the links toward the different libraries implemented for this ticket (TODO: improve description and write unit tests):
https://bitbucket.taktik.be/projects/FLOWR/repos/buffers/browse
https://bitbucket.taktik.be/projects/FLOWR/repos/udp-streamer/browse
https://bitbucket.taktik.be/projects/FLOWR/repos/ts-decryptor/browse

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-10
+v12.14.0
```
Comment (hubjac1) [inline .nvmrc]: I you have no good reason to fix a version. use latest node  12 is better.
`12`

Comment (hubjac1) [reply .nvmrc]: Attention, travis config use node 11

**Comment by hubjac1** [comment]
TODO verify wexond browser

**Review by hubjac1 — APPROVED** [review]

File: src/frontend/src/player.ts [hunk 2]
```diff
@@ -27,252 +63,368 @@ export class Player {
...
-              newStreamData.audio.currentStream = this.currentStreams.audio.currentStream
-              this.currentStreams = newStreamData
-            }
-          } else if (this.currentStreams.audio && this.currentStreams.audio.tracks) {
-            this.currentStreams = newStreamData
-          }
-          this.updateChannelData(this.currentStreams)
+  async stop(): Promise<void> {
+    await this.stopping
+    this.stopping = new Promise(async (resolve) => {
```
Comment (denishamann) [inline src/frontend/src/player.ts]:
```suggestion
    this.stopping = new Promise(async resolve => {
```

File: src/frontend/src/player.ts [hunk 3]
```diff
@@ -27,252 +63,368 @@ export class Player {
...
-    this.playUrl(this.currentStreams.url, this.currentStreams, evt)
+  async retrieveMetadata(url: string): Promise<Ffmpeg.FfprobeData> {
+    let input: string | Readable
 
+    if (this.decryption.use) {
+      input = await this.getDecryptionPipeline(url)
+    } else {
+      input = url
+    }
+
```
Comment (denishamann) [inline src/frontend/src/player.ts:243]: `const input = this.decryption.use ? await this.getDecryptionPipeline(url) : url`

File: src/frontend/src/player.ts [hunk 4]
```diff
@@ -27,252 +63,368 @@ export class Player {
...
+    try {
+      const metadata = await ffprobe(input, { timeout: 30 })
```
Comment (denishamann) [inline src/frontend/src/player.ts]:
```suggestion
      return ffprobe(input, { timeout: 30 })
```

File: src/frontend/src/player.ts [hunk 5]
```diff
@@ -27,252 +63,368 @@ export class Player {
...
+      }
+    }
   }
 
-  playUrl(url: string, streamToPlay: any, evt: any) {
-    console.log('----------- playUrl', url)
+  hasStreamChanged(newStreamData: ICurrentStreams, localCurrentStream: ICurrentStreams | undefined): boolean {
+    // If current and new streams first video track's codec name is different
+    // or if no local current stream
+    // or if audio/subtitle stream does exists anymore
```
Comment (denishamann) [inline src/frontend/src/player.ts]: I don't understand those 3 line comments well

File: src/frontend/src/player.ts (hunk 4 above)
Comment (lmancel) [reply src/frontend/src/player.ts]: I'm trying to use a try / catch block here, if I return a Promise that fails it won't be properly caught with this.

File: src/frontend/src/player.ts (hunk 3 above)
Comment (lmancel) [reply src/frontend/src/player.ts:243]: Sometimes I prefer a simple if / else statement rather than a 100 char long ternary

**Review by denishamann — APPROVED** [review]

**Pull request merged**